proposals/xxxx-experimental-dxil.md

```diff
@@ -0,0 +1,195 @@
+later DXIL version.
+
+## Motivation
+
+During iterative development of the shader compiler it is beneficial to
+implement real lowering into real opcodes to validate that a proposal actually
+solves real world use cases. Traditionally this has been done by marking all in
+development opcodes as expirmental opcodes of the next dxil version release.
+In most cases this is sufficient as the opcodes are accepted into the next
+release but challenges arise when one or more opcodes are rejected or delayed
+from the release while the opcodes following them are not. In the rejection
+case the opcodes must be burned and in the delayed case the opcodes must be
+manually moved to the next release with special casing code. This proposal
+seaks to implemented a systematic method to handle these issues.
+
+## Goals
+This proposal seeks to address the following points:
+ * Needless churn when experimental op are delayed or rejected
+ * Expermental op boundary point is rigid and moves with every SM update
+ * Long term experiments (and potentially extensions) aren't currently feasible
+ * Focused on core api system (hlsl instrinsics and DXIL ops)
+ * Works within the current intrinsics/DXIL op mechanisms
+ * Minimizes overall changes to the system and IHV processes
+ * Straightforward transition route from experimental to stable
+ * IHV drivers can support both experimental and stable versions of an op simultaneously
+   * simplifies migrations
+
+## Non-goals
+Future proposals may address the topics below but this proposal seeks to be a
+smaller isloated change. It intends to solves immediate term challenges
+without investing significant engineering efforts into a generalized solution.
+That said, an attempt is made to avoid proposals that preclude a generalized
+solution. Thus this proposal explicitly avoids addressing these issues:
+ * Full scale generalized extension system
+ * Process development to enable asynchronous non-colliding development
+ * Metadata/RDAT/PSV0/Custom lowering are out of scope for this document
+
+
+
+## Potential DXIL Op Solutions
+
+### Top 1 bit as "is experimental" flag
+
+The top bit of all opcodes is a flag stating if the opcode is experimental.
+
+No structural or shape changes to the DXIL occur, simply the fact that the opcode
+has the high bit set informs that it is experimental. This makes it very easy
+for the compiler and drivers to detect experimental opcodes. When an opcode is
+transistioned to stable the opcode needs to be assigned a stable number.
+This splits the 4 billion opcode space into two 2 billion partions. One for
+stable one for experimental.
+
+This is the simpliest proposal with the least invasive set of changes.
+
+Pros:
+ * Very simple
+ * Quick to implement
+ * Could be implemented "by hand" today by hard coding opcodes
+Cons:
+ * Not a solution for extensions
+ * transistion from experimental to stable isn't just unsetting the bit
+   * other stable ops may have already taken that number
```

Comment:
This line appears to imply that the experimental opcode space isn't independent, because if it was, stable ops would always be in collision as a matter of design.

Basically, this is a con of *any possible* solution, because the only way you have no-op experimental->stable mapping in IR is if you do it the way it's done now, and just put the correct final ops in for the next shader model from the start.

That said, I *do* have a solution that requires no changes to the compiler code that interacts with the ops that ultimately move to final (lowering code etc.), but IR tests still need fixing up somehow.